crates/http-source/src/http_streaming_source.rs

```diff
@@ -0,0 +1,385 @@
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use bytes::BytesMut;
+use encoding_rs::{Encoding, UTF_8};
+use fluvio::Offset;
+use fluvio_connector_common::{tracing::error, Source};
+use futures::{stream::BoxStream, stream::LocalBoxStream, StreamExt};
+use reqwest::{Client, RequestBuilder, Url};
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+use crate::{
+    config::HttpConfig,
+    formatter::{formatter, Formatter, HttpResponseMetadata, HttpResponseRecord},
+};
+
+pub(crate) struct HttpStreamingSource {
+    request: RequestBuilder,
+    delimiter: Vec<u8>,
+    formatter: Arc<dyn Formatter + Sync + Send>,
+    response_metadata: Option<HttpResponseMetadata>,
+    encoding: Option<&'static Encoding>,
+}
+
+#[async_trait]
+impl<'a> Source<'a, String> for HttpStreamingSource {
+    async fn connect(mut self, _offset: Option<Offset>) -> Result<LocalBoxStream<'a, String>> {
+        let request = self
+            .request
+            .try_clone()
+            .context("request must be cloneable")?;
+
+        let response = request.send().await.context("request failed")?;
+
+        self.encoding = Some(transfer_encoding(&response));
```

Comment:
We put `encoding` and `response_metadata` to self and then almost immediately we take it inside the `record_stream`. For that little storage, we are paying by having `Option` and then unwrapping it on every record.
We could instead of setting it on `self` just pass it as arguments to  `record_stream` and further without `Option`.